### ISSUE TITLE
wayland: The assumption that xdg-decoration means SSD is wrong

### ISSUE THREAD

**Issue by PolyMeilex** [libsdl-org/SDL#5653]
This code is not right:
https://github.com/libsdl-org/SDL/blob/d9009388c8924b15f74ec38a26cc62ea42b40110/src/video/wayland/SDL_waylandvideo.c#L895-L897

It was introduced in: https://github.com/libsdl-org/SDL/pull/4559

It assumes that decoration manager being present means that SSD is available. This will only work in compositors that don't support xdg-decoration at all. Compositors that support the protocol and advertise CSD using it will be ignored.

xdg-decoration is a protocol meant for negotiation of CSD/SSD not for SSD exclusively.
An example of that would be a compositor that has a switch in the settings panel that turns SSD on and off. Or a compositor that does not support SSD at all but wants to notify clients about that.

Proper solution is to listen for [zxdg_toplevel_decoration_v1::configure](https://wayland.app/protocols/xdg-decoration-unstable-v1#zxdg_toplevel_decoration_v1:event:configure) event, and update decorations accordingly.

**Comment by flibitijibibo** [comment]
Do we have a compositor that uses this feature? I think we can make this work but being able to test this will be important; it pretty much means we have to short-circuit a major code path _partway through_ (think about it: to even get to this event we have to create the entire window infrastructure, then do a event pump, _then_ we have to tear everything down and completely start over with libdecor) so we'll need a really good real-world test to make sure this handles itself correctly.

**Comment by PolyMeilex** [comment]
We discovered it in [Smithay](https://github.com/Smithay/smithay) project, we have a test compositor called [Anvil](https://github.com/Smithay/smithay/tree/master/anvil) that advertises CSD using xdg-decor. We use it to test if our xdg-decor implementation is correct, by binding SSD/CSD to key combinations and observing if clients get updated accordingly. I can prepare such a test using Anvil if needed, but we don't have binary releases of Anvil so in order to test with it, it has to be built from source.

**Comment by flibitijibibo** [comment]
Understood - for reference, here is where we'll likely need to do this configure check:

https://github.com/libsdl-org/SDL/blob/main/src/video/wayland/SDL_waylandwindow.c#L1259-L1261

After creating the decorations and confirming it wants CSD, we'll have to tear down everything made in this function, load libdecor, then jump back to the top of ShowWindow so that it can take the correct path. Luckily from what I can tell that's _all_ we have to do, since libdecor will at that point take control and handle the configuration events for us.

(For the next revision of the protocol I wonder if making that global would improve the flow of this - AFAIK this would be a global setting anyhow, would it not?)

**Comment by PolyMeilex** [comment]
> After creating the decorations and confirming it wants CSD, we'll have to tear down everything made in this function, load libdecor

That would be a great and relatively simple starting point, as soon as libdecor takes over we get proper dynamic CSD/SSD switching for free. A nitpick would be that if compositor advertises SSD at the startup and changes it to CSD later on, we will ignore that, but that's still better than ignoring it all together. 

Just checked if libdecor handles it, and it works as expected:

https://user-images.githubusercontent.com/20758186/167926874-3697b63b-1d30-41b6-ab18-a8e9364d938a.mp4


So, yeah, as soon as we initialize libdecor we no longer need to worry about xdg-decor details.

**Comment by flibitijibibo** [comment]
See if #5655 makes it work. If not, hopefully the fix from there is really minor...?

**Comment by flibitijibibo** [comment]
One more commit on top, for multi-window: https://github.com/libsdl-org/SDL/commit/6222bd3143073a325db6cea09eb622b976071a5e

(I sure hope somebody tested that PR before merging, because I sure didn't...)

**Comment by PolyMeilex** [comment]
I did:
```bash
git clone https://github.com/libsdl-org/SDL
cd SDL
mkdir build
cd build
../configure
make
```
And I got:
```bash
/home/poly/dev/SDL/src/video/wayland/SDL_waylandwindow.c: In function ‘Wayland_CreateWindow’:
/home/poly/dev/SDL/src/video/wayland/SDL_waylandwindow.c:1909:17: error: ‘struct <anonymous>’ has no member named ‘libdecor’
 1909 |     if (c->shell.libdecor && !IS_POPUP(window)) {
      |                 ^
make: *** [Makefile:745: build/SDL_waylandwindow.lo] Error 1
```
Any ideas what went wrong?

**Comment by flibitijibibo** [comment]
Sort of reads like libdecor wasn't enabled in the configuration (should say whether it found that library or not in the output). Still, a valid error for the non-libdecor path, will fix that now.

**Comment by flibitijibibo** [comment]
Fixed this, sorry it took so long... a bug in something else caught my attention partway through:

https://github.com/libsdl-org/SDL/commit/e4bb3c86496936c1b68141fc4db492426823bbdd

**Comment by PolyMeilex** [comment]
Thanks, in the meanwhile I figured out why it did not load libdecor for me, for some reason it's called libdecoration on my system, that's super weird, but I just built libdecor from source instead, and it works now.

As for SDL I'm still not 100% sure if everything is good. It works in general.

- Case 1
  - XdgDecoration SSD Mode at startup
  - Then change to CSD and back SSD
  - Works like a charm (the window gets remapped, but I understand why that is necessary) 
- Case 2
  - XdgDecoration CSD Mode at startup
  - The window is shown, and instantly hidden (as expected, as it is remapping itself for libdecor)
  - The window never comes back, until mode is switched to SSD and from now on it works as expected.
  - Although I would not assume that SDL did something wrong, the whole operation of initializing toplevel twice with the same surface is kinda exotic, so maybe Smithay is somehow wrong.

So TLDR is that I have to test this with other compositor as well, most likely simple wlroots based one, I don't have any more time today, so I will come back with results tomorrow.

Btw, Thanks you for a quick response and PR!

**Comment by flibitijibibo** [comment]
Sounds good - my guess is it's hanging in HideWindow/ShowWindow somewhere since we're doing some weird stuff in a callback and the surface hasn't presented yet. If gdb doesn't point straight to a hanging line WAYLAND_DEBUG=1 might be interesting.

**Comment by PolyMeilex** [comment]
Ok, I think I figured it out, we simply hit an untested path in Smithay, when the same WlSurface is used twice to create a toplevel initial_configure_sent flag is not resettled, and from the point of view of a compositor it looks like the toplevel was already configured before. That's completely unrelated to SDL of course, and resetting the state when  `NewToplevel` is called did fix everything.

So yeah, it all works like a charm :ok_hand: 
The remapping of a toplevel looks kinda bad, but I understand that libdecor forces you to do this.

> In hindsight, we really should have designed libdecor to be passed a toplevel, having it manage that for us keeps causing major refactors for every change.

I think everyone at Smithay would agree with that, trying to integrate it into our client side toolkit is a nightmare, we even had an idea to maybe pitch something like `unmanaged mode of libdecor` but I don't think anyone is working on that at the moment. I personally just gave up on the idea of using it. Furthermore, I was always under the impression that maybe those problems come from the fact that rust's wayland ecosystem is a lot different then C one, so the API does not translate well, but judging by your comment, you have exactly the same problems as we do. 

Once again, thanks for a quick fix.
